### DISCUSSION TITLE
Converge problem of ImplicitEuler time integrator

### DISCUSSION THREAD

**Discussion by Bearunner** [idaholab/moose#26659] [Q&A General] [answered]
Dear all,

After learning the [Troubleshooting Failed Solves](https://mooseframework.inl.gov/application_usage/failed_solves.html), [Convergence criteria](https://mooseframework.inl.gov/modules/porous_flow/convergence.html) and [Setting Appropriate Convergence Criteria](https://mooseframework.inl.gov/modules/tensor_mechanics/Convergence.html), I have a question about the convergence in MOOSE.

1. Can the default time integrator 'ImplicitEuler ' in MOOSE can solve any kernels/PDE?

2. Whether the [Courant–Friedrichs–Lewy condition](https://en.wikipedia.org/wiki/Courant%E2%80%93Friedrichs%E2%80%93Lewy_condition) **still working for implicit (matrix) solvers**?
**How could we use it in MOOSE?**

 Based on my limited experience, either an increased permeability, or increased injected mass rate at point source or decreased mesh size would require a smaller maximum dt compared to the initial case. **How could we understand this?**

 For example, at the first 50 steps, the hydor-thermal problems in porous_flow can converge by meeting the relative or absolute 
 nonlinear to 1e-8 in the increased permeability case of 3.5e-13 m2. However, it cannot converge at the Time Step 51, time = 547378, dt = 109476 s. The only reason for this non-convergence is the increased permeability from 1e-13 m2 to 3.5e-13 m2. 

 Time Step 50, time = 280256, dt = 56051.9

 0 Nonlinear |R| =  [32m1.053181e-05 
 1 Nonlinear |R| =  [32m3.028068e-08 
 2 Nonlinear |R| =  [32m3.111575e-09 
 Solve Converged! 

 Time Step 51, time = 350321, dt = 70064.9

 0 Nonlinear |R| =  [32m1.037218e-02 
 1 Nonlinear |R| =  [32m1.035678e-02
 2 Nonlinear |R| =  [32m1.023458e-02
 3 Nonlinear |R| =  [32m1.017689e-02 
 4 Nonlinear |R| =  [32m1.004332e-02
 5 Nonlinear |R| =  [32m1.001222e-02 
 6 Nonlinear |R| =  [32m1.001222e-02 
 7 Nonlinear |R| =  [32m1.001222e-02
 8 Nonlinear |R| =  [32m1.001222e-02 
 9 Nonlinear |R| =  [32m1.001222e-02 
 10 Nonlinear |R| =  [32m1.001222e-02

  Nonlinear solve did not converge due to DIVERGED_MAX_IT iterations 10
  Solve Did NOT Converge! 

Thanks for your help.

B.

**Comment by GiudGiud** [comment]
Hello

Yes as long as there are no higher order (second, third etc) time derivatives in the PDE.

There can be some restrictions on the timestep but because it s implicit usually not. 

For your convergence problem, did you try to let the timestepper cut the timestep and try again?

**Reply by GiudGiud** [reply, marked as answer]
The CFL conditions depend on your PDE and on your time integration scheme. They can be derived from a von Neumann stability analysis of your equation iirc

For porous flow it s an advection - diffusion like problem right?

I think the stability of the time integration scheme is not the only thing at play here. There is also the convergence of your Newton / pjfnk nonlinear solve. With a large timestep your initial guess is further from the solution. Unsing a predictor can help

**Reply by Bearunner** [reply]
Thanks for your reply.

1. I am not sure I understand this sentence or not.
> There can be some restrictions on the timestep but because it s  implicit usually not.

Do you mean that 'If it is implicit scheme, for 1st order time derivative, there is no restrictions on the timestep. 
For higher order time derivative, there are restrictions on the timestep.'. Is it right?

2. What is the reason for the unconvergence in the above example since there should be no restriction on the time step beacause I use the implicit first order time derivative? 

3. In addition, with implicit first order time derivative, I found that convergence is dependent on the grid size and flow velocity. For example, high flow velocity results in a small dt.  I am using Newton nonlinear solve and timestepper cutting the timestep.  After 3 steps, the dt increased close to 70064.9, it get cut again.
But from my understanding, any dt should work since I use implicit first order time derivative, which is unconditional stable.

**Reply by GiudGiud** [reply]
1) I did not exactly mean that.
I meant that higher order derivatives are not implemented for 1st order Implicit Euler in MOOSE. I think only the Centered time integrator and Newmark Beta have them implemented. 

> If it is implicit scheme, for 1st order time derivative, there is no restrictions on the timestep

It depends on the time integration scheme, there are more than one implicit scheme.
There are aoways restrictions on the timestep, because you need a certain level of accuracy and because the nonlinear problem solver cannot find a solution if the intiial guess is too far from the solution. 

@cpgr can you please comment on CFL conditions for porous flow?
I dont think it's a classic advective problem right? Because of Darcy's law
